gateway-api: Add required labels and annotations

sayboras opened the pull request:
This commit is to add the required label gateway-name e.g. `gateway.networking.k8s.io/gateway-name`, and propagate all labels and annotations from spec.infrastructure in all generated resources. The main goal is to conform with below GEP.

Relates: https://github.com/kubernetes-sigs/gateway-api/pull/1757

sayboras committed:
gateway-api: Add required labels and annotations

This commit is to add the required label gateway-name e.g.
`gateway.networking.k8s.io/gateway-name`, and propagate all labels and
annotations from spec.infrastructure in all generated resources. The
main goal is to conform with below GEP.

Relates: https://github.com/kubernetes-sigs/gateway-api/pull/1757
Signed-off-by: Tam Mach <tam.mach@cilium.io>

mhofstetter commented on operator/pkg/model/translation/gateway-api/translator.go:
```diff
@@ -21,6 +21,7 @@ var _ translation.Translator = (*gatewayAPITranslator)(nil)
 const (
 	ciliumGatewayPrefix = "cilium-gateway-"
 	owningGatewayLabel  = "io.cilium.gateway/owning-gateway"
```
can we get rid of the Cilium specific label that serves the same purpose? And adapt the internal logic that is making use of it to use the gateway specific label? AFAIK it's only used internally. If it's not possible, we should deprecate it.

sayboras replied:
yes, it's my plan to deprecate internal label in the _next release_, as the controller is using io.cilium.gateway/owning-gateway for filtering logic

sayboras replied:
Thanks, I added the deprecated note for old label.

mhofstetter commented on operator/pkg/model/translation/gateway-api/translator.go:
```diff
@@ -128,11 +137,16 @@ func getService(resource *model.FullyQualifiedResource, allPorts []uint32, label
 		})
 	}
 
+	shortedName := model.Shorten(resource.Name)
```
nit: `shortenedName` or `shortName`

mhofstetter commented on operator/pkg/model/translation/gateway-api/translator.go:
```diff
@@ -110,7 +111,15 @@ func (t *gatewayAPITranslator) Translate(m *model.Model) (*ciliumv2.CiliumEnvoyC
 	}
 
-	return cec, lbSvc, getEndpoints(*source), err
+	shortedName := model.Shorten(source.Name)
```
All these "decorating" aspects (shorten name, adding labels and annotations) to the CEC should be moved into it's own function.

Even if moved to into it's own method, i would avoid having these nested calls to `mergeMap`

sayboras replied:
good point

mhofstetter commented on operator/pkg/model/translation/gateway-api/translator.go:
```diff
@@ -110,7 +111,15 @@ func (t *gatewayAPITranslator) Translate(m *model.Model) (*ciliumv2.CiliumEnvoyC
+		cec.Annotations = mergeMap(cec.Annotations, allAnnotations)
+	}
+
+	return cec, lbSvc, getEndpoints(*source, allLabels, allAnnotations), err
```
for the sake of readability and treating it the same as the Service i would call and assign `getEndpoints(*source, allLabels, allAnnotations)` to a variable next to where we do the same for the service.

mhofstetter commented on operator/pkg/model/translation/gateway-api/translator.go:
```diff
@@ -180,6 +188,39 @@ func getEndpoints(resource model.FullyQualifiedResource) *corev1.Endpoints {
+	if len(res.Labels) == 0 {
+		res.Labels = make(map[string]string)
+	}
+	if len(labels) > 0 {
```
ranging over a nil map is safe. any other reason why we should check before calling `mergeMap`? (here for labels and below for annotations)

sayboras replied:
ah actually, it's to avoid unit test failure, but it seems like can just not initialize allAnnotations and allLabels instead.

mhofstetter commented on operator/pkg/model/translation/gateway-api/translator.go:
```diff
@@ -180,6 +188,39 @@ func getEndpoints(resource model.FullyQualifiedResource) *corev1.Endpoints {
+	}
+
+	shortedName := model.Shorten(resource.Name)
+	if len(res.Labels) == 0 {
```
don't you actually want to test for the `nil` value with `if res.Labels == nil`?

sayboras committed:
gha: Exempt GatewayInfrastructurePropagation Gateway Conformance test

This test is not supported in standard channel, hence we need to add to
the exclude list.

Signed-off-by: Tam Mach <tam.mach@cilium.io>
